linalg/densemat.cpp

```diff
@@ -4392,4 +4393,268 @@ void BatchLUSolve(const DenseTensor &Mlu, const Array<int> &P, Vector &X)
+      }
+
+      // X <- L^{-1} X
+      for (int j = 0; j < m; j++)
+      {
+         const double x_j = x[j];
+         for (int i = j + 1; i < m; i++)
+         {
+            x[i] -= data_all(i, j, e) * x_j;
+         }
+      }
+
+      // X <- U^{-1} X
+      for (int j = m - 1; j >= 0; j--)
+      {
+         const double x_j = (x[j] /= data_all(j, j, e));
+         for (int i = 0; i < j; i++)
+         {
+            x[i] -= data_all(i, j, e) * x_j;
+         }
+      }
+   });
+}
+
+void BatchInverseMatrix(const mfem::Vector &LU,
+                        const int m,
+                        const int NE,
+                        const mfem::Array<int> &P,
+                        mfem::Vector &INV)
+{
+   auto data_all = mfem::Reshape(LU.Read(), m, m, NE);
+   auto piv_all  = mfem::Reshape(P.Read(), m, NE);
+   auto inv_all  = mfem::Reshape(INV.ReadWrite(), m, m, NE);
+
+
+   mfem::forall(NE, [=] MFEM_HOST_DEVICE (int e)
+   {
+      // A^{-1} = U^{-1} L^{-1} P
+      // X <- U^{-1} (set only the upper triangular part of X)
+      double *X          = &inv_all(0, 0, e);
+      double *x          = X;
+      const double *data = &data_all(0, 0, e);
+      const int *ipiv    = &piv_all(0, e);
+
+      for (int k = 0; k < m; k++)
+      {
+         const double minus_x_k = -(x[k] = 1.0 / data[k + k * m]);
+         for (int i = 0; i < k; i++)
+         {
+            x[i] = data[i + k * m] * minus_x_k;
+         }
+         for (int j = k - 1; j >= 0; j--)
+         {
+            const double x_j = (x[j] /= data[j + j * m]);
+            for (int i = 0; i < j; i++)
+            {
+               x[i] -= data[i + j * m] * x_j;
+            }
+         }
+         x += m;
+      }
+
+      // X <- X L^{-1} (use input only from the upper triangular part of X)
+      {
+         int k = m - 1;
+         for (int j = 0; j < k; j++)
+         {
+            const double minus_L_kj = -data[k + j * m];
+            for (int i = 0; i <= j; i++)
+            {
+               X[i + j * m] += X[i + k * m] * minus_L_kj;
+            }
+            for (int i = j + 1; i < m; i++)
+            {
+               X[i + j * m] = X[i + k * m] * minus_L_kj;
+            }
+         }
+      }
+      for (int k = m - 2; k >= 0; k--)
+      {
+         for (int j = 0; j < k; j++)
+         {
+            const double L_kj = data[k + j * m];
+            for (int i = 0; i < m; i++)
+            {
+               X[i + j * m] -= X[i + k * m] * L_kj;
+            }
+         }
+      }
+
+      // X <- X P
+      for (int k = m - 1; k >= 0; k--)
+      {
+         const int piv_k = ipiv[k];
+         if (k != piv_k)
+         {
+            for (int i = 0; i < m; i++)
+            {
+               //Swap<double>(X[i+k*m], X[i+piv_k*m]);
+               mfem::kernels::internal::Swap<double>(X[i + k * m], X[i + piv_k * m]);
+            }
+         }
+      }
+   });
+}
+
+void BatchInverseMatrix(const DenseTensor &LU,
```

Comment:
I think the implementation of this is the same as the one above besides getting `m` and `NE` from the tensor, can you implement one in terms of the other to avoid duplicate code?

Reply:
yes, one thought is we just need to make sure wrap the memory object in the mfem object. There are just in consistencies in using vector and densetensor types to store batches of matrices.

Reply:
you could have a single version that takes the sizes along with c-arrays and just have both call that?